compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/IncrementalJvmCompilerRunner.kt

```diff
@@ -456,24 +470,118 @@ class IncrementalJvmCompilerRunner(
 
     override fun performWorkAfterSuccessfulCompilation(caches: IncrementalJvmCachesManager) {
         if (classpathChanges is ClasspathChanges.ClasspathSnapshotEnabled) {
-            reporter.measure(BuildTime.SAVE_SHRUNK_CURRENT_CLASSPATH_SNAPSHOT_AFTER_COMPILATION) {
+            reporter.measure(BuildTime.SHRINK_AND_SAVE_CURRENT_CLASSPATH_SNAPSHOT_AFTER_COMPILATION) {
                 shrinkAndSaveClasspathSnapshot(classpathChanges, caches.lookupCache)
             }
         }
     }
 
+    /** Used by [shrinkAndSaveClasspathSnapshot]. */
+    private sealed class ShrinkMode {
+        object NoChanges : ShrinkMode()
+
+        class IncrementalNoNewLookups(
+            val shrunkCurrentClasspathAgainstCurrentLookups: List<ClassSnapshotWithHash>,
+        ) : ShrinkMode()
+
+        class Incremental(
+            val currentClasspathSnapshot: List<ClassSnapshotWithHash>,
+            val shrunkCurrentClasspathAgainstPreviousLookups: List<ClassSnapshotWithHash>,
+            val addedOrUpdatedLookupSymbols: Set<LookupSymbolKey>
+        ) : ShrinkMode()
+
+        object NonIncremental : ShrinkMode()
+    }
+
     private fun shrinkAndSaveClasspathSnapshot(classpathChanges: ClasspathChanges.ClasspathSnapshotEnabled, lookupStorage: LookupStorage) {
-        val classpathSnapshot = currentClasspathSnapshot ?: reporter.measure(BuildTime.LOAD_CLASSPATH_SNAPSHOT) {
-            CachedClasspathSnapshotSerializer.load(classpathChanges.classpathSnapshotFiles.currentClasspathEntrySnapshotFiles)
+        // In the following, we'll try to shrink the classpath snapshot incrementally when possible.
+        // For incremental shrinking, we currently use only lookupStorage.addedOrUpdatedLookupSymbols, not
+        // lookupStorage.removedLookupSymbols. It is because:
+        //   1. Tracking removedLookupSymbols is expensive.
```

Comment:
Removed symbols are tracked in the TrackedLookupMap?

Reply:
Right, they are tracked but there are currently no actual calls to the `remove` function in non-test code.

`LookupStorage#removeLookupsFrom` currently doesn't remove entries from the lookup map, probably due to the overhead of doing that.

(Btw, looks like that part of the code may be updated soon to support better task output backup.)

While bullet 1 is still technically correct, I've removed it now to be less confusing. 

> Can you also add a comment that this makes abi non-deterministic to some extent as we may over-keep abi information?

Added a comment.